### ISSUE TITLE
Add option to add files that need to be considered for the cache key in pyproject.toml

### ISSUE THREAD

**Issue by albertferras-vrf** [astral-sh/uv#6255]
Consider the following `pyproject.toml` file, which uses dynamic metadata to reference `requirements.in` files.

```
[project]
name = "myproject"
version = "0.1.0"
requires-python = ">=3.9"
dynamic = ["dependencies"]

[tool.setuptools.dynamic]
dependencies = {file = ["requirements.in"]}
```

When invoking the command `uv pip compile -o requirements.txt pyproject.toml` then setuptools is used to build the project, which is the tool that understands the `dynamic` and `tool.setuptools.dynamic` configuration. Once the build is finished, uv reads the package's `requires.txt` and generates `requirements.txt`.
The problem is that after (only) modifying `requirements.in`, the command `uv pip compile` does not update the `requirements.txt`file. This is because the content `pyproject.toml` did not change and `uv` decide to not reinstall the package (call setuptools) again.
The current workaround is to add `--reinstall` or `--reinstall-package=myproject` to the command or configuration and force setuptools to run.

This was discussed in discord, and the user `@ThiefMaster` suggested that we could add a new configuration under `[tool.uv]` which allows specifying which files have to be considered as part of 'pyproject.toml's cache key. 
One example would be:

```
[tool.uv]
cache_key_files = ["requirements.in", "requirements.dev.in"]
```

Worth mentioning also (not part of this ticket) that it would be good if `uv` supported popular dynamic implementations natively and not need to invoke setuptools for it.

**Comment by charliermarsh** [comment]
I want to work on this but I don't yet have a good solution for `setuptools_scm` or other things that depend on the Git state.

**Comment by charliermarsh** [comment]
I could special-case `.git` in the array though it's not a great API.

**Comment by charliermarsh** [comment]
@konstin @zanieb -- any ideas? If we're gonna support this, I think we need an answer for `setuptools_scm` too since it's so popular.

**Comment by zanieb** [comment]
Well... something like `cache_key_paths = ["path.txt", { file = "other-path.txt }, { git = "." })` is an option? e.g. allow dictionaries so we can define types and default to paths if a simple value is provided?

**Comment by charliermarsh** [comment]
That seems reasonable to me.

**Comment by zanieb** [comment]
How complicated does support for Git state need to be? Like... the commit changed? There are unstaged changes? The tree is dirty? etc.

**Comment by charliermarsh** [comment]
I _think_ "commit hash changed" is fine.

**Comment by charliermarsh** [comment]
I guess `setuptools-scm` will also reflect uncommitted changes, but that may not be necessary.

**Comment by charliermarsh** [comment]
This exists as of v0.4.8. You can do things like:

```toml
[tool.uv]
cache-keys = [{ file = "pyproject.toml" }, { file = "requirements.txt" }, { git = true }]
```
